### DISCUSSION TITLE
Can't find vscode.git built-in extension in WSL

### DISCUSSION THREAD

**Discussion by robertohuertasm** [microsoft/vscode-discussions#219] [Extension Development QnA] [answered]
Hi, I'm working on a extension that makes a small usage of git by leveraging the `vscode.git` built-in [extension](https://github.com/microsoft/vscode/tree/main/extensions/git). 

The problem I have is that **it doesn't seem to be there when using WSL**. I've checked the list of extensions and it's not there. If I switch back to normal windows development, I can see it. 

Anyone knows how to overcome this issue or why is this happening? I haven't found any information about it. I've seen approaches from other extensions that use git (gitlens) and they seem to not be relying on that extension but to get the git path (when the extension is available).

> Note: I posted this in #vscode-extension-dev slack channel and was asked to cross post it here by @digitarald.

Thanks for your help!

**Comment by isidorn** [comment]
For VS Code remote, extensions can run in the remote extension host or in the local one. More details here https://code.visualstudio.com/docs/remote/remote-overview
So it might happen that one of your extensions is running locally and the other remotely.

**Reply by robertohuertasm** [reply]
Thanks @isidorn for your answer. I have still some questions, though.

My extension is declared both as `ui` and `workspace` kind. According to what you say, if my extension is executed locally and the built-it git extension is running remotely (which makes sense), I won't be able to get access to remote extensions at all. That means that from a UI extension I cannot get the api of a remote extension at all. Is that right?

That would explain why `vscode.extensions.getExtension<GitExtension>('vscode.git')` returns `undefined`.

Then, if that's the expected behaviour. Should we assume that `ui` and `workspace` extensions cannot interact with each other?

If that's the case and given that if you declare your extension as supporting both `ui` and `workspace`, vscode will decide where to run it, would it be better to declare the extension only as `workspace` if we want to interact with another `workspace` extension?

Finally, interestingly, when I was debugging the extension (I use WSL while developing), I was able to get access to `vscode.git` from my extension. Would that mean that while debugging `vscode` decided to run my extension in the workspace, but once it is packaged, it decides to use it in the `workspace`?

It was surprising to find this issue while running the extension from the vsix file. While developing and debugging I never found this issue.

Thanks for your help!

**Reply by robertohuertasm** [reply]
One final note. I just looked into [vscode git extension code](https://github.dev/microsoft/vscode/blob/main/extensions/git/package.json) and it seems they don't declare the extension kind. What's the behavior if that is not informed? Will vscode decide? If that's the case, would then it be the same as declaring our extension as supporting both kinds?

**Reply by isidorn** [reply]
@sandy081 might be able to help here.

**Reply by robertohuertasm** [reply]
FWIW, I've just packaged my extension with `workspace` kind only and it worked. Maybe it would be valuable to understand how `vscode` decides where to run an extension that has been declared both as `ui` and `workspace` or that has no `kind` declaration at all.

**Reply by robertohuertasm** [reply]
Another thing I've observed is that if the extension has no `extensionKind` declaration or it just declares the extension as `remote`, users will have the option to install the extension in the remote environment if they want:

![image](https://user-images.githubusercontent.com/696981/200540349-eef71454-e792-4068-9b1b-d1aac97dd11a.png)

On the other hand, if you declare your extension to be both `ui` and `workspace` the user doesn't have this option.

I have this extension installed locally:
![image](https://user-images.githubusercontent.com/696981/200540567-56be7fe7-9318-4972-8c2f-21f88713a9cf.png)

But it's not shown in wsl and I don't have the option to install it manually unless I remove the `ui` declaration:
![image](https://user-images.githubusercontent.com/696981/200540742-8703d9e3-5138-45bf-baeb-ad3b39a8543c.png)

**Comment by sandy081** [comment, marked as answer]
Please go through this doc to understand more about how extension kind works - https://code.visualstudio.com/api/advanced-topics/extension-host. Please let me know if you have any questions there.

**Reply by robertohuertasm** [reply]
Thanks @sandy081! That was very helpful!
